### PR TITLE
[AMORO-2623] Avoid deadlock among TaskRuntime objects

### LINKED ISSUE #2623: [Bug]: Optimizing process stuck after a few failures
### What happened?

If an optimizing task fails constantly (e.g. because of corrupted files), the optimizing task would be stuck at the `RUNNING` state after a few retries, while the task has already failed at the optimizer side.

截屏2024-03-11 16 55 27

### Affects Versions

0.5.1

### What engines are you seeing the problem on?

AMS, Optimizer

### How to reproduce

1. create a table with some data that meets the optimizing conditions.
2. truncate a data file that would take part in the optimizing compaction.
3. AMS tries to optimize the table but fails and retries.
4. after a few retries, AMS would lose track of the task, and leave the task stuck at RUNNING state.

### Relevant log output

_No response_

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by majin1102**
## Why are the changes needed?

Close #2623.

## Brief change log

- avoid deadlock among TaskRuntimes

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by majin1102** [comment]
@link3280

File: amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/optimizing/TaskRuntime.java [hunk 1]
```diff
@@ -105,6 +105,7 @@ public void complete(OptimizerThread thread, OptimizingTaskResult result) {
           token = null;
           threadId = -1;
         });
+    owner.releaseResourcesIfNecessary();
```
Comment (link3280) [inline amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/optimizing/TaskRuntime.java:108]: How does it avoid the deadlock? The deadlock would happen when 2 thread blocks on `owner.acceptResult(this)` before this line.

**Review by link3280 — COMMENTED** [review]
Hi @majin1102 , IIUC the main change is to check the process status and release tasks early right after every completed task (in TaskRuntime), but I don't get how it avoids the deadlock. Please take a look at the comments inline.

**Comment by link3280** [comment]
ping @majin1102

File: amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/optimizing/TaskRuntime.java (hunk 1 above)
Comment (link3280) [reply amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/optimizing/TaskRuntime.java:108]: @rfyu It sounds like what #2684 did.

**Review by XBaith — COMMENTED** [review]
What's the difference between cancelTasks when 2 thread are failed at same time and releaseResourcesIfNecessary

File: amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/optimizing/TaskRuntime.java (hunk 1 above)
Comment (majin1102) [reply amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/optimizing/TaskRuntime.java:108]: > I think two threads are blocked inside `cancelTasks` within `owner.acceptResult(this)`, so wrapping `cancelTask` within `releaseResourcesIfNecessary` and placing it outside `owner.acceptResult(this)` should indeed solve the deadlock issue

Sorry for late replying.

Just like @rfyu said, place canceling All Tasks out of task and process lock could prevent all lock conflict cases involving canceling all tasks.

When two threads steping into owner.acceptResult(), one will wait for lock anyway, I don't quite get your question

**Comment by majin1102** [comment]
> What's the difference between cancelTasks when 2 thread are failed at same time and releaseResourcesIfNecessary

If you mean if  the operation of  canceling all tasks itself would lead to dead lock.

I think the answer is no, concurrent threads would step into task locks in the same order, which would not lead to any dead locks

File: amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/optimizing/TaskRuntime.java (hunk 1 above)
Comment (link3280) [reply amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/optimizing/TaskRuntime.java:108]: I get the idea now. The cancellation is now lock-free. No task lock is needed, thus no deadlock.

**Review by zhoujinsong — APPROVED** [review]
LGTM.

**Pull request merged**